File: java/src/main/java/io/valkey/javabenchmark/command/impl/SetCommand.java
/*
 * Copyright 2025 the original author or authors.
 */
package io.valkey.javabenchmark.command.impl;

import io.valkey.javabenchmark.client.BenchmarkClient;
import io.valkey.javabenchmark.command.Command;
import io.valkey.javabenchmark.config.CommandConfig;
import io.valkey.javabenchmark.engine.KeyGenerator;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ThreadLocalRandom;

/**
 * SET command implementation.
 *
 * @author Ilia Kolominsky
 */
public class SetCommand implements Command {
    private double weight = 1.0;
    private int dataSizeBytes = 256;
    private byte[] valueBuffer;

    @Override
    public String getName() { return "SET"; }

    @Override
    public String getDescription() { return "SET key value"; }

    @Override
    public void configure(CommandConfig config) {
        this.weight = config.getWeight();
        this.dataSizeBytes = config.getDataSizeBytesOrDefault(256);
        this.valueBuffer = new byte[dataSizeBytes];
        ThreadLocalRandom.current().nextBytes(valueBuffer);
    }

    @Override
    public CompletableFuture<CommandResult> execute(BenchmarkClient client, KeyGenerator keyGenerator) {
        byte[] key = keyGenerator.nextKey();
        
        return client.set(key, valueBuffer)
                .thenApply(timedResult -> 
                    CommandResult.success("SET", timedResult.getLatencyMicros()))
                .exceptionally(e -> 
                    // On error, we can't get the real latency from TimedResult, use 0
                    CommandResult.failure("SET", 0, e.getMessage()));
    }

    @Override
    public double getWeight() { return weight; }
}

File: csharp/src/RespBench/Command/Impl/SetCommand.cs
/*
 * Copyright 2025 the original author or authors.
 */
using RespBench.Client;
using RespBench.Config;
using RespBench.Engine;

namespace RespBench.Command.Impl;

public class SetCommand : ICommand
{
    private double _weight = 1.0;
    private byte[] _valueBuffer = Array.Empty<byte>();

    public string Name => "SET";
    public string Description => "SET key value";
    public double Weight => _weight;

    public void Configure(CommandConfig config)
    {
        _weight = config.Weight;
        int dataSize = config.GetDataSizeBytesOrDefault(256);
        _valueBuffer = new byte[dataSize];
        Random.Shared.NextBytes(_valueBuffer);
    }

    public async Task<CommandResult> Execute(IBenchmarkClient client, KeyGenerator keyGenerator)
    {
        byte[] key = keyGenerator.NextKey();
        try
        {
            var result = await client.Set(key, _valueBuffer).ConfigureAwait(false);
            return CommandResult.SuccessResult("SET", result.LatencyMicros);
        }
        catch (Exception e)
        {
            return CommandResult.Failure("SET", 0, e.Message);
        }
    }
}
